src/mpid/ch4/shm/posix/posix_coll_gpu_streamer.h

```diff
@@ -0,0 +1,134 @@
+*/
+
+#ifndef POSIX_COLL_GPU_STREAMER_H_INCLUDED
+#define POSIX_COLL_GPU_STREAMER_H_INCLUDED
+
+#include "../ipc/src/ipc_types.h"
+#include "../ipc/src/ipc_p2p.h"
+#include "../gpu/gpu_post.h"
+#include "../../../include/mpir_err.h"
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_bcast_gpu_stream_read(void *buffer,
+                                                                   MPI_Aint count,
+                                                                   MPI_Datatype datatype,
+                                                                   int root, MPIR_Comm * comm_ptr,
+                                                                   MPIR_Errflag_t errflag)
+{
+    MPIR_FUNC_ENTER;
+    MPIR_CHKLMEM_DECL(1);
+
+    int mpi_errno = MPI_SUCCESS, mpi_errno_ret = MPI_SUCCESS;
+
+    MPI_Aint true_lb;
+    MPIR_Datatype *dt_ptr;
+    bool dt_contig;
+    uintptr_t data_sz;
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, data_sz, dt_ptr, true_lb);
+    /* fallback if datatype is not contiguous or data size is not large enough */
+    if (!dt_contig || data_sz <= MPIR_CVAR_BCAST_STREAM_READ_MSG_SIZE_THRESHOLD) {
+        goto fallback;
+    }
+
+    void *mem_addr = MPIR_get_contig_ptr(buffer, true_lb);
+    int my_rank = MPIR_Comm_rank(comm_ptr);
+    int comm_size = MPIR_Comm_size(comm_ptr);
+
+    MPIDI_IPCI_ipc_attr_t ipc_attr;
+    memset(&ipc_attr, 0, sizeof(ipc_attr));
+    MPIR_GPU_query_pointer_attr(mem_addr, &ipc_attr.gpu_attr);
+
+    /* set up ipc_handles */
+    MPIDI_IPCI_ipc_handle_t *ipc_handles = NULL;
+    MPIR_CHKLMEM_MALLOC(ipc_handles, MPIDI_IPCI_ipc_handle_t *,
+                        sizeof(MPIDI_IPCI_ipc_handle_t) * comm_size, mpi_errno, "IPC handles",
+                        MPL_MEM_COLL);
+
+    MPIDI_IPCI_ipc_handle_t my_ipc_handle;
+    memset(&my_ipc_handle, 0, sizeof(my_ipc_handle));
+    int dev_id = MPL_gpu_get_dev_id_from_attr(&ipc_attr.gpu_attr);
+    /* CPU memory, registered host memory and USM fallback to P2P bcast,
+     * GPU memory will use stream read bcast
+     */
+    if (dev_id >= 0 && ipc_attr.gpu_attr.type == MPL_GPU_POINTER_DEV) {
+        mpi_errno = MPIDI_GPU_get_ipc_attr(mem_addr, my_rank, comm_ptr, &ipc_attr);
+        MPIR_ERR_CHECK(mpi_errno);
+        my_ipc_handle = ipc_attr.ipc_handle;
+    } else {
+        my_ipc_handle.gpu.global_dev_id = -1;
+    }
+
+    /* allgather is needed to exchange all the IPC handles */
+    mpi_errno =
+        MPIR_Allgather_impl(&my_ipc_handle, sizeof(MPIDI_IPCI_ipc_handle_t), MPI_BYTE, ipc_handles,
+                            sizeof(MPIDI_IPCI_ipc_handle_t), MPI_BYTE, comm_ptr, errflag);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    /* check the ipc_handles to make sure all the buffers are on GPU */
+    int errs = 0;
+    for (int i = 0; i < comm_size; i++) {
+        if (ipc_handles[i].gpu.global_dev_id < 0) {
+            errs++;
+            break;
+        }
+    }
+    if (errs > 0) {
+        goto fallback;
+    }
```

Comment:
Every process allocate ipc_handle and use `Allgather` only for this error checking, right? If we add device->host version of local copy, then we only need `Bcast` the root ipc_handle, right?

Reply:
Yes, this Allgather is for the error checking. It adds some overhead compared to Bcast, but we need to make sure all the buffers are on GPU. Also, this algorithm is targeted for large messages, so the added overhead does not affect the overall performance too much. Only messages larger than this CVAR ( MPIR_CVAR_BCAST_STREAM_READ_MSG_SIZE_THRESHOLD) will use the stream read bcast. 

In this algorithm, it is kind of important to keep the root's sendbuf on the GPU so when other ranks copy the send_buf to its own GPU buffer, the communications go through separate Xelinks directly to maximize the bandwidth.